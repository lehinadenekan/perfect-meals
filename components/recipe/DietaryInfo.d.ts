import { DietaryAnalysis } from '@/lib/utils/dietary-classification';
import { Recipe } from '@/lib/types/recipe';

interface DietaryInfoProps {
  analysis: DietaryAnalysis;
  recipe: Recipe;
}

declare const DietaryInfo: React.FC<DietaryInfoProps>;
export default DietaryInfo; 