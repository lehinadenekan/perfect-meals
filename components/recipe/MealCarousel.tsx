'use client';

import React, { useState, useRef, useEffect } from 'react';
import { ChevronLeftIcon, ChevronRightIcon } from '@heroicons/react/24/outline';
import { Recipe } from '@/lib/types/recipe';
import RecipeCard from './RecipeCard';

interface MealCarouselProps {
  title: string;
  recipes: (Recipe & { isFavourite?: boolean })[];
  isLoading?: boolean;
}

const MealCarousel: React.FC<MealCarouselProps> = ({ title, recipes: initialRecipes, isLoading = false }) => {
  const [recipes, setRecipes] = useState(initialRecipes);
  const carouselRef = useRef<HTMLDivElement>(null);
  const [showLeftButton, setShowLeftButton] = useState(false);
  const [showRightButton, setShowRightButton] = useState(true);

  // Card dimensions
  const cardWidth = 240; // width of the card
  const cardGap = 20; // gap between cards
  const scrollAmount = cardWidth + cardGap; // amount to scroll

  // Check scroll position on mount and window resize
  useEffect(() => {
    const checkScrollPosition = () => {
      if (!carouselRef.current) return;

      setShowLeftButton(carouselRef.current.scrollLeft > 0);
      setShowRightButton(
        carouselRef.current.scrollLeft <
        carouselRef.current.scrollWidth - carouselRef.current.clientWidth - 10
      );
    };

    // Initial check
    checkScrollPosition();

    // Add resize listener
    window.addEventListener('resize', checkScrollPosition);
    return () => window.removeEventListener('resize', checkScrollPosition);
  }, [recipes]);

  // Update local state if initialRecipes prop changes
  useEffect(() => {
    setRecipes(initialRecipes);
  }, [initialRecipes]);

  const scroll = (direction: 'left' | 'right') => {
    if (!carouselRef.current) return;

    const currentScroll = carouselRef.current.scrollLeft;
    const newScroll = direction === 'left'
      ? currentScroll - scrollAmount
      : currentScroll + scrollAmount;

    carouselRef.current.scrollTo({
      left: newScroll,
      behavior: 'smooth'
    });

    // Update button visibility after scroll
    setTimeout(() => {
      if (!carouselRef.current) return;
      setShowLeftButton(carouselRef.current.scrollLeft > 0);
      setShowRightButton(
        carouselRef.current.scrollLeft <
        carouselRef.current.scrollWidth - carouselRef.current.clientWidth - 10
      );
    }, 300);
  };

  // --- Callback for Favourite Changes (updates local carousel state) ---
  const handleFavouriteChange = (recipeId: string, newIsFavourite: boolean) => {
    setRecipes(currentRecipes =>
      currentRecipes.map(recipe =>
        recipe.id === recipeId
          ? { ...recipe, isFavourite: newIsFavourite }
          : recipe
      )
    );
    console.log(`Carousel: Recipe ${recipeId} favourite status updated to: ${newIsFavourite}`);
  };

  if (isLoading) {
    return (
      <div className="flex justify-center items-center h-[500px]">
        <div className="animate-spin rounded-full h-12 w-12 border-b-2 border-primary"></div>
      </div>
    );
  }

  if (!recipes.length) {
    return (
      <div className="flex justify-center items-center h-[500px] bg-gray-50 rounded-xl">
        <div className="text-center">
          <p className="text-gray-600 text-lg mb-2">No meals found</p>
          <p className="text-gray-500">Try adjusting your dietary preferences</p>
        </div>
      </div>
    );
  }

  return (
    <div className="mb-8">
      <h2 className="text-2xl font-semibold mb-4 px-4 lg:px-0">{title}</h2>
      <div className="relative">
        {/* Left Navigation Button */}
        {showLeftButton && (
          <button
            onClick={() => scroll('left')}
            className="absolute left-0 top-1/2 -translate-y-1/2 z-10 bg-white/80 hover:bg-white rounded-full p-2 shadow-lg transition-all duration-300 transform hover:scale-110"
            aria-label="Previous recipes"
          >
            <ChevronLeftIcon className="h-6 w-6 text-gray-600" />
          </button>
        )}

        {/* Carousel Container */}
        <div
          ref={carouselRef}
          className="flex gap-5 overflow-x-auto pb-4 scrollbar-hide scroll-smooth"
          style={{
            scrollbarWidth: 'none',
            msOverflowStyle: 'none'
          }}
          onScroll={() => {
            if (!carouselRef.current) return;
            setShowLeftButton(carouselRef.current.scrollLeft > 0);
            setShowRightButton(
              carouselRef.current.scrollLeft <
              carouselRef.current.scrollWidth - carouselRef.current.clientWidth - 10
            );
          }}
        >
          {recipes.map((recipe) => (
            <div
              key={recipe.id}
              className="flex-none animate-fadeIn transition-transform duration-300 w-72"
            >
              <RecipeCard
                recipe={recipe}
                onFavouriteChange={handleFavouriteChange}
              />
            </div>
          ))}
        </div>

        {/* Right Navigation Button */}
        {showRightButton && (
          <button
            onClick={() => scroll('right')}
            className="absolute right-0 top-1/2 -translate-y-1/2 z-10 bg-white/80 hover:bg-white rounded-full p-2 shadow-lg transition-all duration-300 transform hover:scale-110"
            aria-label="Next recipes"
          >
            <ChevronRightIcon className="h-6 w-6 text-gray-600" />
          </button>
        )}
      </div>
    </div>
  );
};

export default MealCarousel; 