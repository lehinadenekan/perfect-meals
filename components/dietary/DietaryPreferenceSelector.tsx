import React, { useState, useEffect, Dispatch, SetStateAction } from 'react';
// Removed Prisma DietType import as it doesn't exist in the schema
// Removed Heroicons imports
import DietSelector from './DietSelector'; // <<<--- !!! THIS ONE STILL NEEDS TO BE FOUND !!!
import RegionSelector from './GeographicFilter'; // <<<--- CORRECTED FILENAME
import { ExcludedFoodsInput } from './ExcludedFoodsInput'; // <<<--- USE NAMED IMPORT
import MealCarousel from '../recipe/MealCarousel'; // <<<--- CORRECTED PATH
import { Recipe } from '@/lib/types/recipe'; // <<<--- CORRECTED IMPORT PATH
import { useSession } from 'next-auth/react';

// Interface for props expected by the component
export interface DietaryPreferenceSelectorProps {
  selectedDiets: string[]; // <<<--- CHANGED DietType[] to string[]
  setSelectedDiets: Dispatch<SetStateAction<string[]>>; // <<<--- CHANGED DietType[] to string[]
  excludedFoods: string[];
  setExcludedFoods: Dispatch<SetStateAction<string[]>>;
  selectedRegions: string[];
  setSelectedRegions: Dispatch<SetStateAction<string[]>>;
  recipes: Recipe[];
  setRecipes: Dispatch<SetStateAction<Recipe[]>>;
  currentStep: number;
  setCurrentStep: Dispatch<SetStateAction<number>>;
}

const DietaryPreferenceSelector: React.FC<DietaryPreferenceSelectorProps> = ({
  selectedDiets,
  setSelectedDiets,
  excludedFoods,
  setExcludedFoods,
  selectedRegions,
  setSelectedRegions,
  recipes,
  setRecipes,
  currentStep,
  setCurrentStep,
}) => {
  const { data: session, status } = useSession();
  const [isSaving, setIsSaving] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  // Effect to save preferences
  useEffect(() => {
    const savePreferences = async () => {
      if (status === 'authenticated' && session?.user?.id) {
        console.log('User authenticated, attempting to save preferences...');
        setIsSaving(true);
        setError(null);

        // Payload uses selectedDiets which is already string[]
        const payload = {
          userId: session.user.id,
          dietTypes: selectedDiets,
          regions: selectedRegions,
          excludedFoods: excludedFoods,
        };
        console.log('Saving preferences payload:', payload);


        try {
          const response = await fetch('/api/user/preferences', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload),
          });

          if (!response.ok) {
            const errorData = await response.json();
            console.error('Failed to save preferences:', errorData);
            setError(errorData.error || 'Failed to save preferences.');
          } else {
             console.log('Preferences saved successfully');
          }

        } catch (err) {
          console.error('Error saving preferences:', err);
          setError('An unexpected error occurred while saving preferences.');
        } finally {
          setIsSaving(false);
        }
      } else {
        console.log('User not authenticated or session data unavailable. Skipping preference save.');
      }
    };

    // Debounce saving preferences
    const handler = setTimeout(() => {
      if (!isLoading) {
        savePreferences();
      }
    }, 1500);

    return () => {
      clearTimeout(handler);
    };
  }, [selectedDiets, selectedRegions, excludedFoods, status, session, isLoading]);

  // Function to fetch recipes based on current selections
  const handleGenerateRecipes = async () => {
    setIsLoading(true);
    setError(null);
    setRecipes([]);

    // Payload uses selectedDiets which is already string[]
    const payload = {
      dietTypes: selectedDiets,
      selectedRegions: selectedRegions,
      excludedFoods: excludedFoods,
    };

    console.log('Generating recipes with payload:', payload);

    try {
      const response = await fetch('/api/recipes/generate', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
      });

      const data = await response.json();

      if (!response.ok || !data.success) {
        const errorMessage = data.message || data.error || 'Failed to generate recipes.';
        console.error('Error generating recipes:', errorMessage, `HTTP status: ${response.status}`);
        setError(errorMessage);
        setRecipes([]);
      } else {
         console.log('Recipes generated successfully:', data.recipes?.length || 0);
         setRecipes(data.recipes || []);
      }
    } catch (err) {
      console.error('Network or other error generating recipes:', err);
      setError('An unexpected error occurred while generating recipes.');
      setRecipes([]);
    } finally {
      setIsLoading(false);
    }
  };

  const handleNext = () => {
    if (currentStep < 3) {
      setCurrentStep(currentStep + 1);
    }
  };

  const handleBack = () => {
    if (currentStep > 1) {
      setCurrentStep(currentStep - 1);
    }
  };

  // const onRemoveRegion = (region: string) => {
  //   setSelectedRegions(prev => prev.filter(r => r !== region));
  // };

  // const onRemoveExcludedFood = (food: string) => {
  //   setExcludedFoods(prev => prev.filter(f => f !== food));
  // };


  const renderStepContent = () => {
    switch (currentStep) {
      case 1:
        return (
          <DietSelector
            selectedPreferences={selectedDiets}
            onChange={setSelectedDiets}
          />
        );
      case 2:
        return (
          <RegionSelector
            selectedRegions={selectedRegions}
            onRegionsChange={setSelectedRegions}
          />
        );
      case 3:
        return (
          <ExcludedFoodsInput
            excludedFoods={excludedFoods}
            onExcludedFoodsChange={setExcludedFoods}
          />
        );
      default:
        return null;
    }
  };

  // No progressValue calculation needed

  return (
    <div
      className="w-full max-w-7xl mx-auto"
      role="region"
      aria-label="dietary preferences selection"
    >
      {/* Progress Bar section is completely removed */}

      {/* Adjusted margin here to provide spacing */}
      <div className="text-center mb-16 mt-8"> {/* Added mt-8 for spacing */}
        <h2 className="text-2xl font-bold">
          {currentStep === 1 && 'Choose Your Dietary Preferences'}
          {currentStep === 2 && 'Select Regional Preferences'}
          {currentStep === 3 && 'Exclude Specific Foods'}
        </h2>
        <p className="text-gray-600 italic mt-2">(optional)</p>
      </div>

      {/* Step Content */}
      {renderStepContent()}

      {/* Navigation Buttons */}
      <div className="flex justify-center gap-4 mt-8">
        {currentStep > 1 && (
          <button
            onClick={handleBack}
            className="px-6 py-3 bg-gray-100 hover:bg-gray-200 text-gray-800 rounded-lg font-semibold transition-colors duration-200"
          >
            Back
          </button>
        )}
        {currentStep < 3 ? (
          <button
            onClick={handleNext}
            className="px-6 py-3 bg-blue-600 hover:bg-blue-700 text-white rounded-lg font-semibold transition-colors duration-200"
          >
            Next
          </button>
        ) : (
          <button
            onClick={handleGenerateRecipes}
            className="px-6 py-3 bg-green-600 hover:bg-green-700 text-white rounded-lg font-semibold transition-colors duration-200"
            disabled={isLoading}
          >
            {isLoading ? 'Loading...' : recipes.length > 0 ? 'Generate New Recipes' : 'Generate Recipes'}
          </button>
        )}
      </div>

      {/* Status Messages */}
      {status === 'unauthenticated' && (
        <p className="text-center mt-4 text-gray-600">
          Log in to save your preferences
        </p>
      )}

      {isSaving && (
        <p className="text-center mt-4 text-gray-600">
          Saving preferences...
        </p>
      )}
       {error && !isLoading && ( // Display error messages only when not loading
        <p className="text-center mt-4 text-red-600">
          Error: {error}
        </p>
      )}


      {/* Recipe Results - Conditionally render based on loading/error state */}
      {!isLoading && !error && recipes.length > 0 && ( // Only show carousel if not loading, no error, and recipes exist
          <div className="mt-8">
            <MealCarousel title="Meals based on your preferences" recipes={recipes} />
          </div>
      )}
      {/* Display message if loading finished, step 3 reached, no error, but no recipes found */}
       {!isLoading && !error && recipes.length === 0 && currentStep === 3 && (
         <p className="text-center mt-8 text-gray-600"> {/* Added more top margin */}
           No recipes found matching your criteria. Try adjusting your preferences.
         </p>
       )}
    </div>
  );
};

export default DietaryPreferenceSelector;