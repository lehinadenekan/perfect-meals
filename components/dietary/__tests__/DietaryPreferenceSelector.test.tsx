import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { axe } from 'jest-axe';
import DietaryPreferenceSelector, { DietaryPreferenceSelectorProps } from '../DietaryPreferenceSelector';
import { server } from '@/src/mocks/server';
import { rest } from 'msw';
import { useSession } from 'next-auth/react';
import { usePreferenceUpdates } from '@/app/hooks/usePreferenceUpdates';
import {
  mockLocalStorage,
  mockSession,
  mockPreferences,
  resetMocks,
  setupLocalStorage,
  DietaryPreferences
} from './test-utils';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import React, { useState } from 'react';
import { DietType } from '@/types/diet';
import { Recipe } from '@/lib/types/recipe';

// Mock next-auth
jest.mock('next-auth/react', () => ({
  useSession: jest.fn(() => mockSession.unauthenticated),
  signIn: jest.fn(),
  signOut: jest.fn(),
}));

// Mock the hooks
jest.mock('@/app/hooks/usePreferenceUpdates', () => ({
  usePreferenceUpdates: jest.fn(() => ({
    updatePreferences: jest.fn()
  }))
}));

// Mock child components
jest.mock('../GeographicFilter', () => ({
  __esModule: true,
  default: jest.fn(() => (
    <div data-testid="mock-geographic-filter">
      <button onClick={() => {}}>Select US</button>
    </div>
  ))
}));

jest.mock('../recipe/MealCarousel', () => ({
  __esModule: true,
  default: jest.fn(() => <div data-testid="mock-meal-carousel" />)
}));

// Setup localStorage mock
Object.defineProperty(window, 'localStorage', { value: mockLocalStorage });

interface PreferenceResponse {
  dietTypes: string[];
  excludedFoods: string[];
  selectedRegions: string[];
  searchInput: string;
}

// Helper component to provide props
const MockProvider: React.FC<Partial<DietaryPreferenceSelectorProps>> = (props) => {
  const [selectedDiets, setSelectedDiets] = useState<DietType[]>(props.selectedDiets || []);
  const [excludedFoods, setExcludedFoods] = useState<string[]>(props.excludedFoods || []);
  const [selectedRegions, setSelectedRegions] = useState<string[]>(props.selectedRegions || []);
  const [recipes, setRecipes] = useState<Recipe[]>(props.recipes || []);
  const [currentStep, setCurrentStep] = useState<number>(props.currentStep || 1);

  const defaultProps: DietaryPreferenceSelectorProps = {
    selectedDiets,
    setSelectedDiets,
    excludedFoods,
    setExcludedFoods,
    selectedRegions,
    setSelectedRegions,
    recipes,
    setRecipes,
    currentStep,
    setCurrentStep,
  };

  return <DietaryPreferenceSelector {...defaultProps} />;
};

describe('DietaryPreferenceSelector', () => {
  beforeAll(() => {
    server.listen();
  });

  beforeEach(() => {
    resetMocks();
    (useSession as jest.Mock).mockClear();
    (usePreferenceUpdates as jest.Mock).mockClear();
  });

  afterEach(() => {
    server.resetHandlers();
  });

  afterAll(() => {
    server.close();
  });

  describe('Accessibility', () => {
    it('should not have any accessibility violations', async () => {
      const { container } = render(<MockProvider />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports keyboard navigation', async () => {
      render(<MockProvider />);
      const buttons = screen.getAllByRole('button');
      
      // Focus first button
      buttons[0].focus();
      expect(document.activeElement).toBe(buttons[0]);

      // Test arrow key navigation
      await userEvent.keyboard('{ArrowRight}');
      expect(document.activeElement).toBe(buttons[1]);
      
      await userEvent.keyboard('{ArrowDown}');
      const bottomRowIndex = buttons.length / 2;
      expect(document.activeElement).toBe(buttons[bottomRowIndex]);
    });
  });

  describe('Loading States', () => {
    it('shows loading state while fetching preferences', async () => {
      render(<MockProvider />);
    });

    it('shows error state when preferences fetch fails', async () => {
      server.use(
        rest.get('http://localhost:3000/api/preferences', (req, res, ctx) => {
          return res(ctx.status(500));
        })
      );

      render(<MockProvider />);
    });
  });

  describe('Preference Management', () => {
    it('loads and displays dietary preferences', async () => {
      setupLocalStorage();
      render(<MockProvider />);

      await waitFor(() => {
        expect(screen.getByText('Vegan')).toBeInTheDocument();
        expect(screen.getByText('Gluten-free')).toBeInTheDocument();
      });
    });

    it('allows selecting dietary preferences', async () => {
      render(<MockProvider />);

      await waitFor(() => {
        const veganButton = screen.getByText('Vegan');
        expect(veganButton).toBeInTheDocument();
        fireEvent.click(veganButton);
        expect(veganButton.closest('[role="button"]')).toHaveClass('border-yellow-400');
      });
    });

    it('saves preferences when user is logged in', async () => {
      (useSession as jest.Mock).mockReturnValue(mockSession.authenticated);

      let savedPreferences: DietaryPreferences | null = null;
      server.use(
        rest.post('http://localhost:3000/api/preferences', async (req, res, ctx) => {
          savedPreferences = await req.json();
          return res(ctx.json({ success: true }));
        })
      );

      render(<MockProvider />);

      await waitFor(() => {
        const veganButton = screen.getByText('Vegan');
        fireEvent.click(veganButton);
      });

      await waitFor(() => {
        expect(savedPreferences).toHaveProperty('dietTypes', ['vegan']);
      });
    });

    it('saves preferences to localStorage when user is not logged in', async () => {
      render(<MockProvider />);

      await waitFor(() => {
        const veganButton = screen.getByText('Vegan');
        fireEvent.click(veganButton);
      });

      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan'])
      );
    });

    it('clears all preferences when clear button is clicked', async () => {
      setupLocalStorage();
      render(<MockProvider />);

      const clearButton = screen.getByRole('button', { name: /clear/i });
      await userEvent.click(clearButton);

      expect(mockLocalStorage.removeItem).toHaveBeenCalledWith('dietary-preferences-selected-diets');
      expect(mockLocalStorage.removeItem).toHaveBeenCalledWith('dietary-preferences-excluded-foods');
      expect(mockLocalStorage.removeItem).toHaveBeenCalledWith('dietary-preferences-selected-regions');
      expect(mockLocalStorage.removeItem).toHaveBeenCalledWith('dietary-preferences-search-input');
    });
  });

  describe('Recipe Generation', () => {
    it('generates recipes based on selected preferences', async () => {
      setupLocalStorage();
      render(<MockProvider />);

      const generateButton = screen.getByRole('button', { name: /generate recipes/i });
      await userEvent.click(generateButton);

      await waitFor(() => {
        expect(screen.getByTestId('mock-meal-carousel')).toBeInTheDocument();
      });
    });

    it('handles recipe generation errors', async () => {
      server.use(
        rest.post('http://localhost:3000/api/recipes/generate', (req, res, ctx) => {
          return res(ctx.status(500));
        })
      );

      render(<MockProvider />);

      const generateButton = screen.getByRole('button', { name: /generate recipes/i });
      await userEvent.click(generateButton);

      await waitFor(() => {
        expect(screen.getByText(/error generating recipes/i)).toBeInTheDocument();
      });
    });
  });

  describe('Preference Combinations', () => {
    it('handles multiple diet selections', async () => {
      render(<MockProvider />);

      const veganButton = screen.getByText('Vegan');
      const glutenFreeButton = screen.getByText('Gluten-free');

      await userEvent.click(veganButton);
      await userEvent.click(glutenFreeButton);

      expect(mockLocalStorage.setItem).toHaveBeenLastCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan', 'gluten-free'])
      );
    });

    it('handles diet and region combinations', async () => {
      render(<MockProvider />);

      const veganButton = screen.getByText('Vegan');
      await userEvent.click(veganButton);

      const regionButton = screen.getByText('Select US');
      await userEvent.click(regionButton);

      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan'])
      );
      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-selected-regions',
        JSON.stringify(['US'])
      );
    });

    it('handles excluded foods with diet types', async () => {
      render(<MockProvider />);
      
      // Select diet type
      const veganButton = screen.getByText('Vegan');
      await userEvent.click(veganButton);

      // Add excluded food
      const excludedFoodInput = screen.getByPlaceholderText(/enter foods to exclude/i);
      await userEvent.type(excludedFoodInput, 'mushrooms{enter}');

      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan'])
      );
      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-excluded-foods',
        JSON.stringify(['mushrooms'])
      );
    });
  });

  describe('Snapshot Testing', () => {
    it('matches snapshot', () => {
      const tree = renderer.create(<MockProvider />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('matches snapshot with different tabs active', () => {
      const treeDiets = renderer.create(<MockProvider />).toJSON();
      expect(treeDiets).toMatchSnapshot();
      const treeExclusions = renderer.create(<MockProvider />).toJSON();
      expect(treeExclusions).toMatchSnapshot();
      const treeRegions = renderer.create(<MockProvider />).toJSON();
      expect(treeRegions).toMatchSnapshot();
    });
  });

  describe('Tab Navigation', () => {
    it('switches tabs correctly', async () => {
      render(<MockProvider />);
      // ... rest of test
    });
  });

  describe('Interaction with Geographic Filter', () => {
    it('updates selected regions when GeographicFilter is used', async () => {
      render(<MockProvider />);
      // ... rest of test
    });
  });

  describe('Concurrent Updates', () => {
    it('handles rapid preference updates correctly', async () => {
      const updatePreferences = jest.fn().mockResolvedValue({ success: true });
      (usePreferenceUpdates as jest.Mock).mockReturnValue({
        updatePreferences
      });

      render(<MockProvider />);

      // Simulate rapid updates
      const veganButton = screen.getByText('Vegan');
      const glutenFreeButton = screen.getByText('Gluten-free');

      await Promise.all([
        userEvent.click(veganButton),
        userEvent.click(glutenFreeButton)
      ]);

      // Should batch the updates
      await waitFor(() => {
        expect(updatePreferences).toHaveBeenCalledTimes(2);
      });
    });

    it('handles concurrent API responses correctly', async () => {
      let requestCount = 0;
      server.use(
        rest.post('http://localhost:3000/api/preferences', async (req, res, ctx) => {
          requestCount++;
          // Simulate varying response times
          await new Promise(resolve => setTimeout(resolve, requestCount * 100));
          return res(ctx.json({ success: true }));
        })
      );

      render(<MockProvider />);

      const veganButton = screen.getByText('Vegan');
      const glutenFreeButton = screen.getByText('Gluten-free');

      // Click buttons in quick succession
      await userEvent.click(veganButton);
      await userEvent.click(glutenFreeButton);

      // Wait for all updates to complete
      await waitFor(() => {
        expect(requestCount).toBe(2);
      });

      // Last state should be correct
      expect(mockLocalStorage.setItem).toHaveBeenLastCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan', 'gluten-free'])
      );
    });

    it('maintains consistency during rapid region changes', async () => {
      render(<MockProvider />);

      // Simulate rapid region selections
      const regionButton = screen.getByText('Select US');
      
      for (let i = 0; i < 5; i++) {
        await userEvent.click(regionButton);
      }

      // Should maintain consistent state
      await waitFor(() => {
        const lastCall = mockLocalStorage.setItem.mock.calls.slice(-1)[0];
        expect(lastCall[0]).toBe('dietary-preferences-selected-regions');
        const regions = JSON.parse(lastCall[1]);
        expect(regions).toHaveLength(1); // Should not have duplicates
        expect(regions).toContain('US');
      });
    });
  });

  describe('Performance', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('debounces preference updates', async () => {
      const updatePreferences = jest.fn().mockResolvedValue({ success: true });
      (usePreferenceUpdates as jest.Mock).mockReturnValue({
        updatePreferences
      });

      render(<MockProvider />);

      // Simulate multiple rapid updates
      const veganButton = screen.getByText('Vegan');
      
      for (let i = 0; i < 5; i++) {
        await userEvent.click(veganButton);
        jest.advanceTimersByTime(100); // Advance time by 100ms
      }

      jest.runAllTimers();

      // Should have debounced the updates
      expect(updatePreferences).toHaveBeenCalledTimes(1);
    });

    it('optimizes recipe generation for large preference sets', async () => {
      setupLocalStorage({
        ...mockPreferences,
        dietTypes: ['vegan', 'gluten-free', 'lactose-free', 'nut-free'],
        excludedFoods: ['mushrooms', 'soy', 'corn', 'wheat'],
        selectedRegions: ['asian', 'mediterranean', 'mexican']
      });

      const startTime = performance.now();
      render(<MockProvider />);
      
      const generateButton = screen.getByRole('button', { name: /generate recipes/i });
      await userEvent.click(generateButton);

      const endTime = performance.now();
      const renderTime = endTime - startTime;

      // Rendering with many preferences should still be fast
      expect(renderTime).toBeLessThan(1000); // 1 second threshold
    });

    it('handles large numbers of excluded foods efficiently', async () => {
      const manyExcludedFoods = Array.from({ length: 100 }, (_, i) => `food-${i}`);
      
      setupLocalStorage({
        ...mockPreferences,
        excludedFoods: manyExcludedFoods
      });

      const startTime = performance.now();
      render(<MockProvider />);
      const endTime = performance.now();

      expect(endTime - startTime).toBeLessThan(500); // 500ms threshold
    });
  });

  describe('API Integration', () => {
    it('integrates with preferences API correctly', async () => {
      const responses: PreferenceResponse[] = [];
      server.use(
        rest.post('http://localhost:3000/api/preferences', async (req, res, ctx) => {
          const body = await req.json();
          responses.push(body);
          return res(ctx.json({ success: true }));
        })
      );

      (useSession as jest.Mock).mockReturnValue(mockSession.authenticated);
      render(<MockProvider />);

      // Make multiple preference changes
      const veganButton = screen.getByText('Vegan');
      const glutenFreeButton = screen.getByText('Gluten-free');
      const excludedFoodInput = screen.getByPlaceholderText(/enter foods to exclude/i);

      await userEvent.click(veganButton);
      await userEvent.click(glutenFreeButton);
      await userEvent.type(excludedFoodInput, 'mushrooms{enter}');

      await waitFor(() => {
        expect(responses).toHaveLength(3);
        expect(responses[responses.length - 1]).toEqual({
          dietTypes: ['vegan', 'gluten-free'],
          excludedFoods: ['mushrooms'],
          selectedRegions: [],
          searchInput: ''
        });
      });
    });

    it('handles API errors gracefully', async () => {
      server.use(
        rest.post('http://localhost:3000/api/preferences', (req, res, ctx) => {
          return res(ctx.status(500), ctx.json({ error: 'Internal Server Error' }));
        })
      );

      (useSession as jest.Mock).mockReturnValue(mockSession.authenticated);
      render(<MockProvider />);

      const veganButton = screen.getByText('Vegan');
      await userEvent.click(veganButton);

      await waitFor(() => {
        expect(screen.getByText(/error saving preferences/i)).toBeInTheDocument();
      });

      // Should fall back to localStorage
      expect(mockLocalStorage.setItem).toHaveBeenCalledWith(
        'dietary-preferences-selected-diets',
        JSON.stringify(['vegan'])
      );
    });

    it('retries failed API calls with exponential backoff', async () => {
      let attempts = 0;
      server.use(
        rest.post('http://localhost:3000/api/preferences', (req, res, ctx) => {
          attempts++;
          if (attempts <= 2) {
            return res(ctx.status(503), ctx.json({ error: 'Service Unavailable' }));
          }
          return res(ctx.json({ success: true }));
        })
      );

      (useSession as jest.Mock).mockReturnValue(mockSession.authenticated);
      render(<MockProvider />);

      const veganButton = screen.getByText('Vegan');
      await userEvent.click(veganButton);

      await waitFor(() => {
        expect(attempts).toBe(3);
      }, { timeout: 5000 });
    });

    it('maintains data consistency across API calls', async () => {
      const savedState: Partial<PreferenceResponse> = {};
      server.use(
        rest.post('http://localhost:3000/api/preferences', async (req, res, ctx) => {
          const body = await req.json();
          Object.assign(savedState, body);
          return res(ctx.json({ success: true }));
        }),
        rest.get('http://localhost:3000/api/preferences', (req, res, ctx) => {
          return res(ctx.json(savedState));
        })
      );

      (useSession as jest.Mock).mockReturnValue(mockSession.authenticated);
      const { rerender } = render(<MockProvider />);

      // Make changes
      const veganButton = screen.getByText('Vegan');
      await userEvent.click(veganButton);

      // Unmount and remount to test persistence
      rerender(<></>);
      rerender(<MockProvider />);

      await waitFor(() => {
        const veganButtonAfterRerender = screen.getByText('Vegan');
        expect(veganButtonAfterRerender.closest('[role="button"]')).toHaveClass('border-yellow-400');
      });
    });
  });
});