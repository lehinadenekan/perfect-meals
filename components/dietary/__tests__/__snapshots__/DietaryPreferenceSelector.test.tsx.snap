// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DietaryPreferenceSelector Snapshot Testing matches snapshot 1`] = `null`;

exports[`DietaryPreferenceSelector Snapshot Testing matches snapshot with different tabs active 1`] = `null`;

exports[`DietaryPreferenceSelector Snapshot Testing matches snapshot with different tabs active 2`] = `null`;

exports[`DietaryPreferenceSelector Snapshot Testing matches snapshot with different tabs active 3`] = `null`;
